File: simplerecyclerviewadapter/src/main/java/io/github/horaciocome1/simplerecyclerviewadapter/SimpleRecyclerViewAdapter.java
package io.github.horaciocome1.simplerecyclerviewadapter;

import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;

import java.util.List;

import androidx.annotation.NonNull;
import androidx.recyclerview.widget.RecyclerView;

import static io.github.horaciocome1.simplerecyclerviewadapter.Constants.ITEM_LAYOUT_NOT_SPECIFIED;

public class SimpleRecyclerViewAdapter extends RecyclerView.Adapter<SimpleRecyclerViewAdapter.ViewHolder> {

    private List list;
    private int itemLayout;
    private OnBindViewHolder onBindViewHolder;
    private ViewHolder viewHolder = null;

    private SimpleRecyclerViewAdapter(List list, int itemLayout, OnBindViewHolder onBindViewHolder) {
        this.list = list;
        this.itemLayout = itemLayout;
        this.onBindViewHolder = onBindViewHolder;
    }

    @NonNull
    @Override
    public ViewHolder onCreateViewHolder(@NonNull ViewGroup parent, int viewType) {
        if (!(itemLayout == ITEM_LAYOUT_NOT_SPECIFIED)) {
            View view = LayoutInflater.from(parent.getContext()).inflate(itemLayout, parent, false);
            viewHolder = new ViewHolder(view);
        }
        return viewHolder;
    }

    @Override
    public void onBindViewHolder(@NonNull ViewHolder holder, int position) {
        if (!(onBindViewHolder == null)) onBindViewHolder.onBind(holder, list.get(position));
    }

    @Override
    public int getItemCount() {
        if (list == null)
            return 0;
        return list.size();
    }

    class ViewHolder extends RecyclerView.ViewHolder {

        private View view;

        private ViewHolder(@NonNull View itemView) {
            super(itemView);
            view = itemView;
        }

        public View getViewById(int viewId) {
            return view.findViewById(viewId);
        }

    }

    public interface OnBindViewHolder {

        void onBind(@NonNull ViewHolder holder, Object object);

    }

    public static class Builder {

        private List list = null;
        private int itemLayout = ITEM_LAYOUT_NOT_SPECIFIED;
        private SimpleRecyclerViewAdapter.OnBindViewHolder onBindViewHolder = null;

        public Builder setList(List list) {
            this.list = list;
            return this;
        }

        public Builder setItemLayout(int itemLayout) {
            this.itemLayout = itemLayout;
            return this;
        }

        public Builder setOnBindViewHolder(SimpleRecyclerViewAdapter.OnBindViewHolder onBindViewHolder) {
            this.onBindViewHolder = onBindViewHolder;
            return this;
        }

        public SimpleRecyclerViewAdapter build() {
            return new SimpleRecyclerViewAdapter(list, itemLayout, onBindViewHolder);
        }

    }


}

File: simplerecyclerviewadapter/src/main/java/io/github/horaciocome1/simplerecyclerviewadapter/SimpleRecyclerViewAdapter.kt
/*
 * Copyright 2018 Horácio Flávio Comé Júnior
 *
 *    Licensed under the Apache License, Version 2.0 (the "License");
 *    you may not use this file except in compliance with the License.
 *    You may obtain a copy of the License at
 *
 *        http://www.apache.org/licenses/LICENSE-2.0
 *
 *    Unless required by applicable law or agreed to in writing, software
 *    distributed under the License is distributed on an "AS IS" BASIS,
 *    WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 *    See the License for the specific language governing permissions and
 *    limitations under the License.
 */

package io.github.horaciocome1.simplerecyclerviewadapter

import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import androidx.recyclerview.widget.RecyclerView
import androidx.recyclerview.widget.RecyclerView.Adapter

const val NO_ITEM_LAYOUT = -1

class SimpleRecyclerViewAdapter<DataType> : Adapter<SimpleRecyclerViewAdapter.ViewHolder>() {

    var itemLayout: Int = NO_ITEM_LAYOUT
    var list = ArrayList<DataType>()
    private var onBind = { _: ViewHolder, _: DataType -> Unit}

    override fun onCreateViewHolder(parent: ViewGroup, viewType: Int): ViewHolder {
        val view = LayoutInflater.from(parent.context).inflate(itemLayout, parent, false)
        return ViewHolder(view)
    }

    override fun getItemCount(): Int { return list.size }

    override fun onBindViewHolder(holder: ViewHolder, position: Int) { onBind(holder, list[position]) }

    fun setOnBindViewHolder(onBind: (SimpleRecyclerViewAdapter.ViewHolder, DataType) -> Unit) { this.onBind = onBind }

    fun addItem(item: DataType) {
        list.add(item)
        notifyItemInserted(list.size)
    }

    fun setItem(item: DataType, position: Int) {
        list[position] = item
        notifyItemChanged(position)
    }

    fun removeItem(position: Int) {
        list.removeAt(position)
        notifyItemRemoved(position)
    }

    fun restoreItem(item: DataType, position: Int) {
        list.add(position, item)
        notifyItemInserted(position)
    }

    class ViewHolder(view: View) : RecyclerView.ViewHolder(view)

}